### PR TITLE
core, all: split vm.Context into BlockContext and TxContext

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
As you can see, this is a mixed bag, ring tx are way faster, while other benchmarks are a bit slower

This PR:
```
BenchmarkInsertChain_empty_memdb-8            	    4575	    306455 ns/op	  178274 B/op	     510 allocs/op
BenchmarkInsertChain_empty_diskdb-8           	    3570	    351276 ns/op	  182059 B/op	     532 allocs/op
BenchmarkInsertChain_valueTx_memdb-8          	    3625	    373037 ns/op	  203530 B/op	     684 allocs/op
BenchmarkInsertChain_valueTx_diskdb-8         	    3140	    376081 ns/op	  207843 B/op	     705 allocs/op
BenchmarkInsertChain_valueTx_100kB_memdb-8    	    1000	   1521589 ns/op	  746068 B/op	     684 allocs/op
BenchmarkInsertChain_valueTx_100kB_diskdb-8   	     758	   1472932 ns/op	 1056727 B/op	     709 allocs/op
BenchmarkInsertChain_uncles_memdb-8           	    3351	    498745 ns/op	  195669 B/op	     689 allocs/op
BenchmarkInsertChain_uncles_diskdb-8          	    2412	    489038 ns/op	  201543 B/op	     707 allocs/op
BenchmarkInsertChain_ring200_memdb-8          	     146	   8301682 ns/op	 4108689 B/op	   30131 allocs/op
BenchmarkInsertChain_ring200_diskdb-8         	     144	   8333911 ns/op	 4153333 B/op	   29473 allocs/op
BenchmarkInsertChain_ring1000_memdb-8         	     120	   9980914 ns/op	 4597596 B/op	   34056 allocs/op
BenchmarkInsertChain_ring1000_diskdb-8        	     100	  10021250 ns/op	 4659023 B/op	   33422 allocs/op
```
Master:
```
BenchmarkInsertChain_empty_memdb-8            	    5017	    217500 ns/op	  177494 B/op	     505 allocs/op
BenchmarkInsertChain_empty_diskdb-8           	    5956	    221143 ns/op	  180055 B/op	     524 allocs/op
BenchmarkInsertChain_valueTx_memdb-8          	    5241	    274443 ns/op	  203068 B/op	     699 allocs/op
BenchmarkInsertChain_valueTx_diskdb-8         	    5948	    271572 ns/op	  206797 B/op	     717 allocs/op
BenchmarkInsertChain_valueTx_100kB_memdb-8    	    1767	   1085632 ns/op	  741137 B/op	     698 allocs/op
BenchmarkInsertChain_valueTx_100kB_diskdb-8   	    1005	   1071035 ns/op	 1064578 B/op	     724 allocs/op
BenchmarkInsertChain_uncles_memdb-8           	    3931	    354559 ns/op	  194458 B/op	     683 allocs/op
BenchmarkInsertChain_uncles_diskdb-8          	    4341	    385686 ns/op	  199261 B/op	     703 allocs/op
BenchmarkInsertChain_ring200_memdb-8          	     120	  10077005 ns/op	 4594716 B/op	   35317 allocs/op
BenchmarkInsertChain_ring200_diskdb-8         	      97	  11685285 ns/op	 4646687 B/op	   34663 allocs/op
BenchmarkInsertChain_ring1000_memdb-8         	      86	  13789912 ns/op	 5194921 B/op	   39543 allocs/op
BenchmarkInsertChain_ring1000_diskdb-8        	      94	  14345609 ns/op	 5252341 B/op	   38873 allocs/op
```

### CODE REVIEW

File: core/state_processor.go [hunk 1]
```diff
@@ -81,20 +86,12 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
...
-		return nil, err
-	}
+func applyTransaction(msg types.Message, config *params.ChainConfig, bc ChainContext, author *common.Address, gp *GasPool, statedb *state.StateDB, header *types.Header, tx *types.Transaction, usedGas *uint64, cfg vm.Config, blockContext vm.BlockContext) (*types.Receipt, error) {
 	// Create a new context to be used in the EVM environment
-	context := NewEVMContext(msg, header, bc, author)
+	txContext := NewEVMTxContext(msg)
 	// Create a new environment which holds all relevant information
 	// about the transaction and calling mechanisms.
-	vmenv := vm.NewEVM(context, statedb, config, cfg)
+	vmenv := vm.NewEVM(blockContext, txContext, statedb, config, cfg)
```
Comment (holiman) [inline core/state_processor.go:94]: Do you really need to instantiate a new evm here? Shouldn't it suffice to do that at block boundaries

Comment (MariusVanDerWijden) [reply core/state_processor.go:94]: I don't know, the comment on NewEVM says "The returned EVM is not thread safe and should only ever be used *once*."
I could of course add a _Reset_ method of some kind which would probably greatly improve things.

Comment (MariusVanDerWijden) [reply core/state_processor.go:94]: Adding that change resulted in the following test times:
```
BenchmarkInsertChain_empty_memdb-8            	    4536	    262351 ns/op	  185415 B/op	     516 allocs/op
BenchmarkInsertChain_empty_diskdb-8           	    6140	    293717 ns/op	  187502 B/op	     536 allocs/op
BenchmarkInsertChain_valueTx_memdb-8          	    3690	    338649 ns/op	  202926 B/op	     682 allocs/op
BenchmarkInsertChain_valueTx_diskdb-8         	    4546	    324259 ns/op	  206864 B/op	     705 allocs/op
BenchmarkInsertChain_valueTx_100kB_memdb-8    	    1296	   1124334 ns/op	  743574 B/op	     687 allocs/op
BenchmarkInsertChain_valueTx_100kB_diskdb-8   	    1038	   1306500 ns/op	 1059309 B/op	     714 allocs/op
BenchmarkInsertChain_uncles_memdb-8           	    3290	    441751 ns/op	  202184 B/op	     695 allocs/op
BenchmarkInsertChain_uncles_diskdb-8          	    3888	    462218 ns/op	  206863 B/op	     715 allocs/op
BenchmarkInsertChain_ring200_memdb-8          	     189	   6440408 ns/op	 2505123 B/op	   28874 allocs/op
BenchmarkInsertChain_ring200_diskdb-8         	     183	   6861615 ns/op	 2556595 B/op	   28222 allocs/op
BenchmarkInsertChain_ring1000_memdb-8         	     152	   7822861 ns/op	 3117208 B/op	   33133 allocs/op
BenchmarkInsertChain_ring1000_diskdb-8        	     148	   8386961 ns/op	 3207706 B/op	   32458 allocs/op
```
The empty inserts are still a bit slower, but the ring inserts are ~40% faster

**Comment by MariusVanDerWijden** [comment]
Adding these changes on top of the passing vm.Context by pointer produces the following results:
```

BenchmarkInsertChain_empty_memdb-8            	    6682	    184174 ns/op	  177967 B/op	     518 allocs/op
BenchmarkInsertChain_empty_diskdb-8           	    7000	    227932 ns/op	  180654 B/op	     537 allocs/op
BenchmarkInsertChain_valueTx_memdb-8          	    6634	    209066 ns/op	  190173 B/op	     697 allocs/op
BenchmarkInsertChain_valueTx_diskdb-8         	    6020	    237992 ns/op	  193433 B/op	     715 allocs/op
BenchmarkInsertChain_valueTx_100kB_memdb-8    	    1734	    785544 ns/op	  728628 B/op	     697 allocs/op
BenchmarkInsertChain_valueTx_100kB_diskdb-8   	    1501	   1020587 ns/op	 1041337 B/op	     734 allocs/op
BenchmarkInsertChain_uncles_memdb-8           	    5558	    272765 ns/op	  193894 B/op	     697 allocs/op
BenchmarkInsertChain_uncles_diskdb-8          	    5248	    315847 ns/op	  199513 B/op	     716 allocs/op
BenchmarkInsertChain_ring200_memdb-8          	     183	   6714158 ns/op	 2987597 B/op	   32318 allocs/op
BenchmarkInsertChain_ring200_diskdb-8         	     177	   6954004 ns/op	 3048000 B/op	   31668 allocs/op
BenchmarkInsertChain_ring1000_memdb-8         	     147	   8271707 ns/op	 3637109 B/op	   36795 allocs/op
BenchmarkInsertChain_ring1000_diskdb-8        	     142	   8579421 ns/op	 3729780 B/op	   36126 allocs/op
```

File: core/state_transition.go [hunk 2]
```diff
@@ -230,8 +230,8 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 	}
 	msg := st.msg
 	sender := vm.AccountRef(msg.From())
-	homestead := st.evm.ChainConfig().IsHomestead(st.evm.BlockNumber)
-	istanbul := st.evm.ChainConfig().IsIstanbul(st.evm.BlockNumber)
+	homestead := st.evm.ChainConfig().IsHomestead(st.evm.Context.BlockNumber)
```
Comment (holiman) [inline core/state_transition.go:233]: The PR contains a lot of these changes -- might be better to leave it as is, to make the change smaller?

Comment (holiman) [reply core/state_transition.go:233]: Edit: ignore that. You made the Context explicit now, so the change is needed

**Comment by holiman** [comment]
Screenshot_2020-10-15 Dual Geth - Grafana
This is taken after a fast-sync, about 4 hours after both machines reached the head. This PR running on `bench03` (right), against master on `bench04` (left). The `execution` appears to be a few milliseconds faster on this PR. 
It might be a good idea to swap the machines in a while, and see if the same effect it still seen.

**Comment by holiman** [comment]
Here's 3.5 days of block processing. This PR on the right-hand side

Screenshot_2020-10-18 Dual Geth - Grafana

Zooming in on `execution`
Screenshot_2020-10-18 Dual Geth - Grafana(1)

**Comment by holiman** [comment]
Last six hours with them swapped: 

Screenshot_2020-10-19 Dual Geth - Grafana
Screenshot_2020-10-19 Dual Geth - Grafana(1)

**Comment by holiman** [comment]
PR on `bench03`, master on `bench04`: `41.5ms` vs `45.1ms` == PR is `8%` faster on `execution`
PR on `bench04`, mater on `bench03`: `50.2ms` vs `56.1ms` == PR is `11%` faster on `execution`. 
So the 'real' number might be somwhere around 9.5% faster execution, shaving off `~5ms` in total block processing, out of `200ms` (overall `2.5%` faster )

**Review by holiman — APPROVED** [review]

File: core/evm.go [hunk 3]
```diff
@@ -35,26 +35,31 @@ type ChainContext interface {
...
 		Coinbase:    beneficiary,
 		BlockNumber: new(big.Int).Set(header.Number),
 		Time:        new(big.Int).SetUint64(header.Time),
 		Difficulty:  new(big.Int).Set(header.Difficulty),
 		GasLimit:    header.GasLimit,
-		GasPrice:    new(big.Int).Set(msg.GasPrice()),
+	}
+}
+
+func NewEVMTxContext(msg Message) vm.TxContext {
```
Comment (karalabe) [inline core/evm.go:59]: dox ;P

File: core/vm/evm.go [hunk 4]
```diff
@@ -99,6 +95,12 @@ type Context struct {
 	Difficulty  *big.Int       // Provides information for DIFFICULTY
 }
 
+type TxContext struct {
```
Comment (karalabe) [inline core/vm/evm.go:98]: docs pls

**Pull request merged**